File: apps/web/src/types/allocator.ts
export interface ShadowEpisode {
  index: number;
  variant: string;
  reward: number;
  candidate_profit: number;
  baseline_profit: number;
  guardrail_violated: boolean;
  realised_roas: Record<string, number>;
  disabled_after_episode: boolean;
  safety_override: boolean;
}

export interface ShadowValidationCheck {
  name: string;
  status: boolean;
  value: number;
  threshold: number;
  observed_baseline_runs?: number;
  required_baseline_runs?: number;
}

export interface ShadowValidationSummary {
  episodes: number;
  safety_override_rate: number;
  disabled_variants: string[];
}

export interface ShadowValidationStressEpisode {
  index: number;
  variant: string;
  guardrail_violated: boolean;
  disabled_after_episode: boolean;
}

export interface ShadowValidationStressConfig {
  episodes: number;
  epsilon: number;
  seed: number;
  max_guardrail_breaches: number;
}

export interface ShadowValidationStressTest {
  config: ShadowValidationStressConfig;
  guardrail_violations: number;
  guardrail_breach_counts: Record<string, number>;
  selection_counts: Record<string, number>;
  disabled_variants: string[];
  episodes: ShadowValidationStressEpisode[];
  assertions: Record<string, boolean>;
}

export interface ShadowValidation {
  checks: ShadowValidationCheck[];
  summary: ShadowValidationSummary;
  notes: string[];
  stress_test: ShadowValidationStressTest;
}

export interface ShadowRunReport {
  generated_at: string;
  average_reward: number;
  guardrail_violations: number;
  q_values: Record<string, number>;
  selection_counts: Record<string, number>;
  episodes: ShadowEpisode[];
  guardrail_breach_counts: Record<string, number>;
  disabled_variants: string[];
  diagnostics: Record<string, number>;
  config: Record<string, unknown>;
  scenario: Record<string, unknown>;
  validation: ShadowValidation;
}

export interface SaturationMarket {
  name: string;
  allocated_spend: number;
  share: number;
  fair_share: number;
  min_share: number;
  revenue: number;
  roas: number;
  saturation_ratio: number;
  lift_vs_current: number;
  current_spend: number;
  weather_multiplier: number;
  guardrail_binding: boolean;
}

export interface SaturationSummary {
  profit: number;
  baseline_profit: number;
  profit_lift: number;
  weighted_fairness_gap: number;
  max_fairness_gap: number;
  total_revenue: number;
  total_spend: number;
}

export interface SaturationReport {
  generated_at: string;
  total_budget: number;
  fairness_floor: number;
  roas_floor: number;
  summary: SaturationSummary;
  markets: SaturationMarket[];
  allocator: Record<string, unknown>;
}

File: shared/schemas/allocator.py
from __future__ import annotations

from datetime import datetime
from typing import Dict

from pydantic import BaseModel, Field


class ShadowEpisode(BaseModel):
    index: int
    variant: str
    reward: float
    candidate_profit: float
    baseline_profit: float
    guardrail_violated: bool
    realised_roas: Dict[str, float]
    disabled_after_episode: bool
    safety_override: bool


class ShadowValidationCheck(BaseModel):
    name: str
    status: bool
    value: float
    threshold: float
    observed_baseline_runs: int | None = None
    required_baseline_runs: int | None = None


class ShadowValidationSummary(BaseModel):
    episodes: int
    safety_override_rate: float
    disabled_variants: list[str]


class ShadowValidationStressEpisode(BaseModel):
    index: int
    variant: str
    guardrail_violated: bool
    disabled_after_episode: bool


class ShadowValidationStressConfig(BaseModel):
    episodes: int
    epsilon: float
    seed: int
    max_guardrail_breaches: int


class ShadowValidationStressTest(BaseModel):
    config: ShadowValidationStressConfig
    guardrail_violations: int
    guardrail_breach_counts: Dict[str, int]
    selection_counts: Dict[str, int]
    disabled_variants: list[str]
    episodes: list[ShadowValidationStressEpisode]
    assertions: Dict[str, bool]


class ShadowValidation(BaseModel):
    checks: list[ShadowValidationCheck]
    summary: ShadowValidationSummary
    notes: list[str]
    stress_test: ShadowValidationStressTest


class ShadowRunReport(BaseModel):
    generated_at: datetime = Field(default_factory=datetime.utcnow)
    average_reward: float
    guardrail_violations: int
    q_values: Dict[str, float | None]
    selection_counts: Dict[str, int]
    episodes: list[ShadowEpisode]
    guardrail_breach_counts: Dict[str, int]
    disabled_variants: list[str]
    diagnostics: Dict[str, float]
    config: Dict[str, object]
    scenario: Dict[str, object]
    validation: ShadowValidation


class SaturationMarket(BaseModel):
    name: str
    allocated_spend: float
    share: float
    fair_share: float
    min_share: float
    revenue: float
    roas: float
    saturation_ratio: float
    lift_vs_current: float | None
    current_spend: float
    weather_multiplier: float
    guardrail_binding: bool
    fairness_gap: float
    fairness_ratio: float | None = None
    target_spend: float
    spend_delta_vs_target: float
    floor_shortfall: float


class SaturationSummary(BaseModel):
    profit: float
    baseline_profit: float
    profit_lift: float
    weighted_fairness_gap: float
    max_fairness_gap: float
    total_revenue: float
    total_spend: float
    normalized_fairness_gap: float
    under_allocated_markets: int
    total_floor_shortfall: float
    max_floor_shortfall: float


class SaturationReport(BaseModel):
    generated_at: datetime = Field(default_factory=datetime.utcnow)
    total_budget: float
    fairness_floor: float
    roas_floor: float
    summary: SaturationSummary
    markets: list[SaturationMarket]
    allocator: Dict[str, object]


__all__ = [
    "SaturationMarket",
    "SaturationReport",
    "SaturationSummary",
    "ShadowEpisode",
    "ShadowRunReport",
    "ShadowValidation",
    "ShadowValidationCheck",
    "ShadowValidationStressConfig",
    "ShadowValidationStressEpisode",
    "ShadowValidationStressTest",
    "ShadowValidationSummary",
]
